lib/jobs/drive-partition-control.js

```diff
@@ -0,0 +1,222 @@
+                return true;
+            }
+            if (parser) {
+                var result = parser.call(null, task.stdout);
+                if (result.error) {
+                    logger.error('The parser tells command execution has failure on node.', {
+                        nodeId: self.nodeId,
+                        responseObject: task
+                    });
+                    return true;
+                }
+                parsedResult.push(result);
+            }
+            else {
+                parsedResult.push(task.stdout);
+            }
+            return false;
+        });
+
+        if (failed) {
+            return { error: true };
+        }
+        else {
+            return { error: false, data: parsedResult };
+        }
+    };
+
+    /**
+     * The common function to remotely run commands in microkernel and handle its response,
+     * This function is wrapped as Promise
+     * @param {Array} cmds - Array of String, the batched commands that need be executed
+     * @param {Function} [parser] - The function to parse the command response, this can be omitted.
+     * @param {Integer} [timeout] - The timeout (in milliseconds) to wait the command response,
+     * default is 60 sec.
+     * @return {Promise}
+     */
+    DrivePartitionJob.prototype.runRemoteCommand = function(cmds, parser, timeout) {
+        var self = this;
+        timeout = timeout || 60*1000; //default set 1 min as timeout
+        return new Promise(function(resolve, reject) {
+            self._subscribeRequestCommands(function() {
+                logger.debug("Received command request from node. Sending commands.", {
+                    nodeId: self.nodeId,
+                    commands: cmds
+                });
+                return {
+                    identifier: self.nodeId,
+                    tasks: cmds
+                };
+            });
+
+             self._subscribeRespondCommands(function(data) {
+                var result = self.handleResponse(data, parser);
+                if (result.error) {
+                    reject(new Error('The command response shows some error'));
+                }
+                else {
+                    resolve(result.data);
+                }
+            });
+        }).timeout(timeout, 'Timeout to wait for the node response');
+    };
+
+    /**
+     * @memberOf DrivePartitionJob
+     * @function
+     * @return {Promise}
+     */
+    DrivePartitionJob.prototype._run = function() {
+        var self = this;
+
+        // Called if this job is used to render a script template
+        self._subscribeRequestProperties(function() {
+            return self.options;
+        });
+
+        //if not disks are specified, then do nothing
+        if (!self.options.driveIds || self.options.driveIds.length === 0) {
+            logger.debug('no driveIds is specified, so ignore drive partition operation.');
+            self._done();
+            return Promise.resolve();
```

Comment:
I think @yyscamper's intent here is so the _run function can be tested more easily, however I still think it's better to do `return job.deferred.promise` in test because it's more flexible for _run functions.